Add E2E test to test checkout performance

harriswong opened the pull request:
#### Changes proposed in this Pull Request
We would like to create an E2E test to measure metrics such as first paint, first contentful paint, etc. The test will run 3 times to take an average. It measures 3 things on the `/checkout` page: 
- Stripe element checkout
- UPE checkout
- WooPay checkout

The PR uses a similar technique as woocommerce-blocks https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tests/utils/performance.ts. 

#### Testing instructions
1. Follow E2E instructions here(https://github.com/Automattic/woocommerce-payments/tree/develop/tests/e2e) to setup.
2. Make sure your `tests/e2e/deps/wcp-dev-tools` folder is up-to-date. You can do `git pull` to make sure it is the latest.
3. Run `npm run test:e2e-performance`
4. You should see the test passed and a new file is created under `tests/e2e/reports/checkout-performance.txt`. Each line of the file is a JSON record of a performance metric. It should look something like this:
```
{"description":"Stripe element: Average","serverResponse":3693.6966666641333,"firstPaint":76.56499991814296,"domContentLoaded":89.84833319361012,"loaded":372.0683332843085,"firstContentfulPaint":76.56499991814296,"firstBlock":2955.526666715741}
{"description":"Stripe UPE: Average","serverResponse":4223.398333260168,"firstPaint":90.24666668847203,"domContentLoaded":123.69833327829838,"loaded":465.96499998122454,"firstContentfulPaint":90.24666668847203,"firstBlock":3817.1183334973953}
{"description":"WooPay: Average","serverResponse":4126.750000131627,"firstPaint":77.12333323433995,"domContentLoaded":95.77999993537863,"loaded":338.5383333079517,"firstContentfulPaint":77.12333323433995,"firstBlock":3170.6916666589677}
```

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] 'QA Testing Not Applicable'
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

harriswong commented on tests/e2e/utils/performance.js:
```diff
@@ -0,0 +1,37 @@
+export async function getLoadingDurations() {
```
Similar to the function from woocommerce-blocks https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tests/utils/performance.ts. The differences are that I can't destruct the `performance.getEntriesByType(..)` assignment or use array prototype.find. I get the following errors:
`ReferenceError: _slicedToArray2` 
`ReferenceError: _find`

It could be a puppeteer problem since these are used within `page.evaulate()`. Our puppeteer version is 2 and it relies on [puppeteer-utils](https://github.com/Automattic/puppeteer-utils/blob/master/package.json#L26). Blocks uses [jest-environment-puppeteer](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/package.json#L178) where its puppeteer is at 14. It doesn't seem to have these issues. I think it would be good to upgrade it to the newer puppeteer version.

In any case, I worked around the `ReferenceError` and this function works in this branch.

harriswong commented:
👋  @achyuthajoy @htdat I am adding an E2E test to measure `/checkout` performance. I want to check in with you both since you have reviewed some of the E2E PRs. Please let me know if I should ping someone else. Thanks for your time!

harriswong commented:
> Is there a specific reason why the results are stored as a file instead of outputting to the terminal?

No. There isn't any specific reason to store it in a file. As long as we can get the metrics in our local dev/CI, then that should be enough for our needs. I saw woocommerce-blocks' PR (https://github.com/woocommerce/woocommerce-blocks/pull/6119) which they stored in a file for output. I thought that was a pretty good idea and so I took that direction as well.

> Is there plans to run these tests periodically via Github actions? If yes, then storing results in file might not be ideal.

I think it will be good to run this in Github Actions in the future, but maybe not for every PR. I was thinking it could be run manually from a trigger phrase in the comment, or per release candidate. In this case, would rendering output to the terminal be a better approach?

harriswong commented:
👋  @achyuthajoy, I want to check in just In case you missed my reply above. Thanks!

achyuthajoy commented:
Hey @harriswong 👋! Here is what we've discussed on slack today:

* It would be better to add the performance test as a new group under the existing E2E test suite. This makes it easier to manage in case there are any changes to the test suite.
* The new test group won't be a part of the CI/CD pipeline because we specify the test groups to be tested. Right now it's `wcpay`, `subscriptions` & `blocks`.

achyuthajoy commented:
> I saw woocommerce-blocks' PR (https://github.com/woocommerce/woocommerce-blocks/pull/6119) which they stored in a file for output. I thought that was a pretty good idea and so I took that direction as well.

GitHub actions deletes artifacts uploaded after a specific period. So if you wish to have historic data, then it's best to log it to the console, which will be preserver. Or you can send the data to another service to get it recorded.

> I think it will be good to run this in Github Actions in the future, but maybe not for every PR. I was thinking it could be run manually from a trigger phrase in the comment, or per release candidate

Yes, instead of running it on every PR, you can create a workflow which can be triggered manually, by comment or when a specific tag/branch is created.

harriswong commented:
@achyuthajoy I reverted the commit that introduced the `"test:e2e-performance": "NODE_CONFIG_DIR='tests/e2e/config' wp-scripts test-e2e --config tests/e2e/config/jest.performance.config.js",` command. Running `npm run test:e2e` will run the `tests/e2e/specs/performance/payment-methods.spec.js` though. Is it beneficial to use a technique similar to https://github.com/Automattic/woocommerce-payments/blob/a661754d6def0fbcf9be9113adcbe8eb66852075/tests/e2e/utils/flows.js#L45-L52 so that local dev can skip the performance tests?

harriswong commented:
> > I saw woocommerce-blocks' PR ([woocommerce/woocommerce-blocks#6119](https://github.com/woocommerce/woocommerce-blocks/pull/6119)) which they stored in a file for output. I thought that was a pretty good idea and so I took that direction as well.
> 
> GitHub actions deletes artifacts uploaded after a specific period. So if you wish to have historic data, then it's best to log it to the console, which will be preserver. Or you can send the data to another service to get it recorded.

Ok thanks! For now, I don't think we are running it in CI and we don't need the historic data. If this changes, I will pursue a different approach so that we can preserve the results. Let me know what you think.

achyuthajoy commented:
> I reverted the commit that introduced the `"test:e2e-performance": "NODE_CONFIG_DIR='tests/e2e/config' wp-scripts test-e2e --config tests/e2e/config/jest.performance.config.js",` command. Running `npm run test:e2e` will run the `tests/e2e/specs/performance/payment-methods.spec.js` though. Is it beneficial to use a technique similar to
> 
> https://github.com/Automattic/woocommerce-payments/blob/a661754d6def0fbcf9be9113adcbe8eb66852075/tests/e2e/utils/flows.js#L45-L52
> so that local dev can skip the performance tests?
> 
> **updates** I added a commit([b4ca77c](https://github.com/Automattic/woocommerce-payments/commit/b4ca77c36a88a01423369955ecc95ef949f60d9d)) to introduce the `SKIP_WC_PERFORMANCE_TESTS` env variable. Let me know if this is undesired, I can revert the commit.

@harriswong It would be much cleaner if we keep a separate jest config for performance tests. That way, you don't need to exclude the test using environment variables and keep the NPM script to run just the performance test. 

What are your thoughts?

harriswong commented:
> > I reverted the commit that introduced the `"test:e2e-performance": "NODE_CONFIG_DIR='tests/e2e/config' wp-scripts test-e2e --config tests/e2e/config/jest.performance.config.js",` command. Running `npm run test:e2e` will run the `tests/e2e/specs/performance/payment-methods.spec.js` though. Is it beneficial to use a technique similar to
> > https://github.com/Automattic/woocommerce-payments/blob/a661754d6def0fbcf9be9113adcbe8eb66852075/tests/e2e/utils/flows.js#L45-L52
> > 
> > so that local dev can skip the performance tests?
> > **updates** I added a commit([b4ca77c](https://github.com/Automattic/woocommerce-payments/commit/b4ca77c36a88a01423369955ecc95ef949f60d9d)) to introduce the `SKIP_WC_PERFORMANCE_TESTS` env variable. Let me know if this is undesired, I can revert the commit.
> 
> @harriswong It would be much cleaner if we keep a separate jest config for performance tests. That way, you don't need to exclude the test using environment variables and keep the NPM script to run just the performance test.
> 
> What are your thoughts?

@achyuthajoy To clarify, do you mean it is better to keep this commit https://github.com/Automattic/woocommerce-payments/pull/4344/commits/a326f98434c61118023132e2d8a3723ab85d5812?

achyuthajoy commented:
> To clarify, do you mean it is better to keep this commit [a326f98](https://github.com/Automattic/woocommerce-payments/commit/a326f98434c61118023132e2d8a3723ab85d5812)

@harriswong Yes. That would be best to keep the performance test separate from current E2E. 

Was this commit removed based on our previous discussion? If yes, I'm so sorry, I might've missed something while checking the implementation.

harriswong commented:
> > > > I saw woocommerce-blocks' PR ([woocommerce/woocommerce-blocks#6119](https://github.com/woocommerce/woocommerce-blocks/pull/6119)) which they stored in a file for output. I thought that was a pretty good idea and so I took that direction as well.
> > > 
> > > 
> > > GitHub actions deletes artifacts uploaded after a specific period. So if you wish to have historic data, then it's best to log it to the console, which will be preserver. Or you can send the data to another service to get it recorded.
> > 
> > 
> > Ok thanks! For now, I don't think we are running it in CI and we don't need the historic data. If this changes, I will pursue a different approach so that we can preserve the results. Let me know what you think.
> 
> Sure. If historic data is not needed, writing to file or outputting to terminal works. I also noticed that terminal logs are also removed after a while.
> 
> @harriswong Do you plan to set up a Github workflow for running performance tests periodically/with manual trigger? Or just keep this for local testing.
> 
> For more accurate results I would suggest setting up a GitHub actions workflow. It's because running locally on mac with docker can be very slow and might not be the original performance.

@achyuthajoy Thanks for the suggestion! I wasn't planning on setting up a GitHub action workflow as of yet. I was planning to run this locally so that we can measure a before-and-after effect on this other PR(https://github.com/Automattic/woocommerce-payments/pull/3968). We also don't know if we will be going forward with https://github.com/Automattic/woocommerce-payments/pull/3968. But you are right that the results may vary from machine to machine. Maybe creating a manual trigger is better. Even though https://github.com/Automattic/woocommerce-payments/pull/3968 is in draft, I can see some benefits in triggering it manually in the PR's comment. This way, the metric will be more consistent compared to each of our individual machines. I will look into adding a github action to this. 

Can this be a follow-up or would you prefer to have these all in the same PR?

achyuthajoy commented:
> Can this be a follow-up or would you prefer to have these all in the same PR?

The GH workflow can be another PR.

harriswong commented:
> > To clarify, do you mean it is better to keep this commit [a326f98](https://github.com/Automattic/woocommerce-payments/commit/a326f98434c61118023132e2d8a3723ab85d5812)
> 
> @harriswong Yes. That would be best to keep the performance test separate from current E2E.
> 
> Was this commit removed based on our previous discussion? If yes, I'm so sorry, I might've missed something while checking the implementation.

No problem at all! `git` makes these super easy to do. Thanks for taking the time to review! 🙇 I do appreciate the feedback on this. Ok, I will bring back the https://github.com/Automattic/woocommerce-payments/commit/a326f98434c61118023132e2d8a3723ab85d5812 and revert the commit that introduces `SKIP_WC_PERFORMANCE_TESTS`. 👍

harriswong commented:
> > Can this be a follow-up or would you prefer to have these all in the same PR?
> 
> The GH workflow can be another PR.

Added follow up here: https://github.com/Automattic/woocommerce-payments/issues/4389

harriswong commented:
> No problem at all! `git` makes these super easy to do. Thanks for taking the time to review! 🙇 I do appreciate the feedback on this. Ok, I will bring back the [a326f98](https://github.com/Automattic/woocommerce-payments/commit/a326f98434c61118023132e2d8a3723ab85d5812) and revert the commit that introduces `SKIP_WC_PERFORMANCE_TESTS`. 👍

Done. Running `npm run test:e2e-performance` should print something like this:
![image](https://user-images.githubusercontent.com/572862/176798766-2a8fc46a-d231-4cf2-8734-fa1cc6b5503f.png)

@achyuthajoy please take another look.

achyuthajoy reviewed (changes requested):
@harriswong I've tested the changes locally and can confirm that it's working a expected. Awesome work on this one.

I've left a few comments. Once it's cleared, it's good to merge 👍

harriswong commented on tests/e2e/utils/flows.js:
```diff
@@ -490,4 +490,62 @@ export const merchantWCP = {
 		}
 	},
+
+	activateWooPay: async () => {
```
cli would be great! It will make it a lot faster! Thanks for the note.

harriswong commented on tests/e2e/specs/performance/payment-methods.spec.js:
```diff
@@ -0,0 +1,115 @@
+		recreatePerformanceFile();
+	} );
+
+	describe( 'Stripe element', () => {
```
The reason I did this was because I wanted to have a clean file per performance run. At the same time, I wanted each scenario (Stripe element, UPE, WooPay without UPE, etc) to have its own setup. As a result, I did 
```
describe // first level to always wipe the file before we start a run
     describe // Stripe element setup
     describe // UPE setup
     describe // WooPay without UPE setup
```

Looking at this again, I wonder if I can mix `beforeAll` and `beforeEach` with 1 layer of `describe`, such that wiping the files execute only once. 👀

achyuthajoy reviewed (changes requested):
@harriswong I noticed that the report file created is not added to `gitignore`. Can you add it?

Also, tested the changes on a linux machine against a local server instance. The results are much faster that running on a Mac.

![Screenshot from 2022-07-08 12-46-10](https://user-images.githubusercontent.com/1161872/177939064-8193df03-c570-4b88-9078-efb882d4d216.png)

harriswong commented:
Hi @achyuthajoy , I merged the latest `develop` and added a line to the `.gitignore`. Please take a look again!

> Also, tested the changes on a linux machine against a local server instance. The results are much faster that running on a Mac.

Thanks for giving this a test! The power of linux 💪

achyuthajoy reviewed (approved):
LGTM 🚢 

Thanks for your awesome work on this 👍

harriswong commented:
@achyuthajoy The "PHP testing (7.1)" job is stuck but this recent PR https://github.com/Automattic/woocommerce-payments/pull/4425/files removed 7.1

Should we ignore this stuck test and merge?

achyuthajoy commented:
@harriswong Yes, please ignore the 7.1 test. It was removed when WP & WC versions were bumped as part of the next release
